### PR TITLE
techdebt: [CDS-37837]: Serverless First Deployment Logic For Rollback

### PR DESCRIPTION
**Pull request by piyush-bhuwalka**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/33712)

### CODE REVIEW

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

File: 125-cd-nextgen/src/main/java/io/harness/cdng/serverless/ServerlessAwsLambdaRollbackStep.java [hunk 1]
```diff
@@ -108,10 +108,12 @@ public TaskRequest obtainTaskAfterRbac(
         (ServerlessAwsLambdaRollbackDataOutcome) serverlessRollbackDataOptionalOutput.getOutput();
     ServerlessGitFetchOutcome serverlessGitFetchOutcome =
         (ServerlessGitFetchOutcome) serverlessGitFetchOptionalOutput.getOutput();
-    if (EmptyPredicate.isEmpty(rollbackDataOutcome.getPreviousVersionTimeStamp())) {
+    if (rollbackDataOutcome.isFirstDeployment() == false
```
Comment (sainath-b) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/serverless/ServerlessAwsLambdaRollbackStep.java:111]: !rollbackDataOutcome.isFirstDeployment()

File: 930-delegate-tasks/src/main/java/io/harness/delegate/serverless/ServerlessAwsLambdaDeployCommandTaskHandler.java [hunk 2]
```diff
@@ -79,6 +79,7 @@ public class ServerlessAwsLambdaDeployCommandTaskHandler extends ServerlessComma
   private ServerlessAwsLambdaInfraConfig serverlessAwsLambdaInfraConfig;
   private long timeoutInMillis;
   private String previousDeployTimeStamp;
+  private Boolean firstDeployment;
```
Comment (sainath-b) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/serverless/ServerlessAwsLambdaDeployCommandTaskHandler.java:82]: Change this to boolean?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/serverless/ServerlessAwsLambdaDeployCommandTaskHandler.java [hunk 3]
```diff
@@ -211,6 +213,13 @@ private void configureCredential(ServerlessDeployRequest serverlessDeployRequest
   private void prepareRollbackData(ServerlessDeployRequest serverlessDeployRequest, LogCallback executionLogCallback,
       ServerlessDelegateTaskParams serverlessDelegateTaskParams) throws Exception {
     executionLogCallback.saveExecutionLog(format("Preparing Rollback Data..%n%n"));
+    if (!serverlessAwsCommandTaskHelper.cloudFormationTemplateExists(
```
Comment (sainath-b) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/serverless/ServerlessAwsLambdaDeployCommandTaskHandler.java:216]: function name shall be cloudFormationStackExists

File: 930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsCFHelperServiceDelegateImpl.java [hunk 4]
```diff
@@ -101,6 +104,32 @@ public List<AwsCFTemplateParamsData> getParamsData(AwsInternalConfig awsConfig,
     return emptyList();
   }
 
+  @Override
+  public Boolean stackExists(AwsInternalConfig awsConfig, String region, String stackId) {
```
Comment (sainath-b) [inline 930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsCFHelperServiceDelegateImpl.java:108]: boolean

File: 125-cd-nextgen/src/main/java/io/harness/cdng/serverless/ServerlessAwsLambdaRollbackStep.java [hunk 5]
```diff
@@ -108,10 +108,12 @@ public TaskRequest obtainTaskAfterRbac(
...
+        && EmptyPredicate.isEmpty(rollbackDataOutcome.getPreviousVersionTimeStamp())) {
       return TaskRequest.newBuilder()
-          .setSkipTaskRequest(
-              SkipTaskRequest.newBuilder().setMessage("No previous version exist. Skipping rollback.").build())
+          .setSkipTaskRequest(SkipTaskRequest.newBuilder()
+                                  .setMessage("Serverless Aws Lambda Deploy step was not executed. Skipping rollback.")
```
Comment (sainath-b) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/serverless/ServerlessAwsLambdaRollbackStep.java:115]: No previous version exist. Skipping rollback.
This message suits here even after changes. Let me know what you think about it. @piyush-bhuwalka

File: 125-cd-nextgen/src/main/java/io/harness/cdng/serverless/ServerlessAwsLambdaRollbackStep.java (hunk 1 above)
Comment (piyush-bhuwalka) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/serverless/ServerlessAwsLambdaRollbackStep.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/serverless/ServerlessAwsLambdaDeployCommandTaskHandler.java (hunk 2 above)
Comment (piyush-bhuwalka) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/serverless/ServerlessAwsLambdaDeployCommandTaskHandler.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/serverless/ServerlessAwsLambdaDeployCommandTaskHandler.java (hunk 3 above)
Comment (piyush-bhuwalka) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/serverless/ServerlessAwsLambdaDeployCommandTaskHandler.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/serverless/ServerlessAwsLambdaRollbackCommandTaskHandler.java [hunk 6]
```diff
@@ -177,18 +179,46 @@ private ServerlessRollbackResponse rollback(ServerlessRollbackRequest serverless
...
+              executionLogCallback, serverlessRollbackRequest, serverlessRollbackRequest.getManifestContent())) {
+        executionLogCallback.saveExecutionLog(
+            format("Nothing to Remove..%n"), LogLevel.INFO, CommandExecutionStatus.SUCCESS);
+        serverlessRollbackResponseBuilder.commandExecutionStatus(CommandExecutionStatus.SUCCESS);
+        serverlessRollbackResponseBuilder.serverlessRollbackResult(serverlessAwsLambdaRollbackResultBuilder.build());
+        return serverlessRollbackResponseBuilder.build();
+      } else {
+        response = serverlessAwsCommandTaskHelper.remove(serverlessClient, serverlessDelegateTaskParams,
+            executionLogCallback, timeoutInMillis, serverlessManifestConfig, serverlessAwsLambdaInfraConfig);
+        addTimestampToResponse = false;
```
[the remark this answers is not in the document]
Comment (piyush-bhuwalka) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/serverless/ServerlessAwsLambdaRollbackCommandTaskHandler.java]: Done

File: 930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsCFHelperServiceDelegateImpl.java (hunk 4 above)
Comment (piyush-bhuwalka) [reply 930-delegate-tasks/src/main/java/software/wings/service/impl/aws/delegate/AwsCFHelperServiceDelegateImpl.java]: Done

File: 125-cd-nextgen/src/main/java/io/harness/cdng/serverless/ServerlessAwsLambdaRollbackStep.java (hunk 5 above)
Comment (piyush-bhuwalka) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/serverless/ServerlessAwsLambdaRollbackStep.java]: Yes that should also make sense. Reverted to the previous message

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Pull request closed without merging**